File: src/SolTake.Core/Multimedia.cs
﻿using Newtonsoft.Json;

namespace SolTake.Core
{
    public class Multimedia
    {
        public string ContainerName { get; private set; }
        public string BlobName { get; private set; }
        public string? BlobNameOfFrame { get; private set; }
        public long Size { get; private set; }
        public double Height { get; private set; }
        public double Width { get; private set; }
        public double Duration { get; private set; }
        public MultimediaType MultimediaType { get; private set; }

        [JsonConstructor]
        private Multimedia(string containerName, string blobName, string? blobNameOfFrame, long size, double height, double width, double duration, MultimediaType multimediaType)
        {
            ContainerName = containerName;
            BlobName = blobName;
            BlobNameOfFrame = blobNameOfFrame;
            Size = size;
            Height = height;
            Width = width;
            Duration = duration;
            MultimediaType = multimediaType;
        }

        private Multimedia(string containerName, string blobName, long size, double height, double width)
        {
            ContainerName = containerName;
            BlobName = blobName;
            Size = size;
            Height = height;
            Width = width;
            Duration = 0;
        }

        private Multimedia(string containerName, string blobName, string blobNameOfFrame, long size, double height, double width, double duration)
        {
            ContainerName = containerName;
            BlobName = blobName;
            BlobNameOfFrame = blobNameOfFrame;
            Size = size;
            Height = height;
            Width = width;
            Duration = duration;
        }

        public static Multimedia CreateImage(string containerName, string blobName, long size, double height, double width)
            => new(containerName, blobName, size, height, width) { MultimediaType = MultimediaType.Image };

        public static Multimedia CreateVideo(string containerName, string blobName, string blobNameOfFrame, long size, double height, double width, double duration)
            => new(containerName, blobName, blobNameOfFrame, size, height, width, duration) { MultimediaType = MultimediaType.Video };


        public static Multimedia Clone(Multimedia media) => new(
            media.ContainerName,
            media.BlobName,
            media.BlobNameOfFrame,
            media.Size,
            media.Height,
            media.Width,
            media.Duration,
            media.MultimediaType
        );

    }
}

File: MobileApp/my_social_app/lib/custom_packages/media/models/multimedia.dart
import 'package:flutter/cupertino.dart';
import 'package:json_annotation/json_annotation.dart';
import 'package:my_social_app/custom_packages/media/models/dimention.dart';
import 'package:my_social_app/custom_packages/media/models/media.dart';
import 'package:my_social_app/custom_packages/media/models/multimedia_type.dart';
import 'package:my_social_app/custom_packages/media/models/remote_image.dart';
import 'package:my_social_app/custom_packages/media/models/remote_video.dart';
part 'multimedia.g.dart';

@JsonSerializable()
@immutable
class Multimedia{
  final String containerName;
  final String blobName;
  final String? blobNameOfFrame;
  final int size;
  final double height;
  final double width;
  final double duration;
  final MultimediaType multimediaType;

  double get aspectRatio => width / height;
  
  const Multimedia({
    required this.containerName,
    required this.blobName,
    required this.blobNameOfFrame,
    required this.size,
    required this.height,
    required this.width,
    required this.duration,
    required this.multimediaType
  });

  factory Multimedia.fromJson(Map<String, dynamic> json) => _$MultimediaFromJson(json);
  Map<String, dynamic> toJson() => _$MultimediaToJson(this);

  Media toMedia() => 
    multimediaType == MultimediaType.image
      ? RemoteImage(
          blobName: blobName,
          containerName: containerName,
          dimention: Dimention(width: width, height: height),
          size: size
        )
      : RemoteVideo(
          blobName: blobName,
          containerName: containerName,
          dimention: Dimention(width: width, height: height),
          size: size,
          blobNameOfFrame: blobNameOfFrame!,
          duration: duration
        );
}